cmd/tealdbg/localLedger.go

```diff
@@ -138,8 +161,35 @@ func makeBalancesAdapter(
 		}
 	}
 
+	assetsExist := make(map[basics.AssetIndex]bool, len(assets))
+	for _, aid := range assets {
+		for _, ad := range balances {
+			if _, ok := ad.AssetParams[aid]; ok {
+				assetsExist[aid] = true
+			}
+		}
+	}
+
 	// painless mode creates all missed global states and opt-in all mentioned accounts
 	if painless {
+		for _, aid := range assets {
+			if ok := assetsExist[aid]; !ok {
+				// create balance record and AppParams for this asset
+				addr, err := getRandomAddress()
+				if err != nil {
+					return nil, AppState{}, err
+				}
+				ad := basics.AccountData{
+					AssetParams: map[basics.AssetIndex]basics.AssetParams{
+						aid: makeAssetParams(addr),
+					},
+					Assets: map[basics.AssetIndex]basics.AssetHolding{
+						aid: makeAssetHolding(),
+					},
+				}
+				balances[addr] = ad
+			}
+		}
```

Comment:
why does it needed at all if these assets are already in `balances`?

Reply:
If indexer is not being used then all the referenced assets in the transaction may not exist. If the painless option is enabled, this code generates a balance record for each asset that doesn't exist.

Reply:
ah, sorry, overlooked `!ok`